src/vmm/src/devices/virtio/queue.rs

```diff
@@ -488,13 +522,40 @@ impl Queue {
     fn set_avail_event<M: GuestMemory>(&mut self, val: u16, mem: &M) {
         debug_assert!(self.is_layout_valid(mem));
 
+        // Used ring has layout:
+        // struct UsedRing {
+        //     flags: u16,
+        //     idx: u16,
+        //     ring: [UsedElement; <queue size>],
+        //     avail_event: u16,
+        // }
+        // We calculate offset into `avail_event` field.
+        let avail_event_offset = std::mem::size_of::<u16>()
+            + std::mem::size_of::<u16>()
+            + std::mem::size_of::<UsedElement>() * usize::from(self.actual_size());
         let avail_event_addr = self
             .used_ring
-            .unchecked_add(u64::from(4 + 8 * self.actual_size()));
+            .unchecked_add(usize_to_u64(avail_event_offset));
 
         mem.write_obj(val, avail_event_addr).unwrap();
     }
 
+    fn set_next_used<M: GuestMemory>(&mut self, val: u16, mem: &M) {
```

Comment:
nit: Instead of `val: u16`, use `next_used_idx: u16` to make the name a bit more descriptive.

Reply:
done